WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldsListViewModel.swift

```diff
@@ -0,0 +1,97 @@
+    @Published private(set) var combinedList: [CustomFieldUI] = []
+
+    @Published private var editedFields: [CustomFieldUI] = []
+    @Published private var addedFields: [CustomFieldUI] = []
+    var hasChanges: Bool {
+        !editedFields.isEmpty || !addedFields.isEmpty
+    }
+
+    init(customFields: [CustomFieldViewModel]) {
+        self.originalCustomFields = customFields
+        updateCombinedList()
+    }
+}
+
+// MARK: - Items actions
+extension CustomFieldsListViewModel {
+    func editField(at index: Int, newField: CustomFieldUI) {
+        guard index >= 0 && index < combinedList.count else {
+            DDLogError("⛔️ Error: Invalid index for editing a custom field")
+            return
+        }
+
+        let oldField = combinedList[index]
+
+        if newField.id == nil {
+            // If there's no id, it means we're now editing a newly added field.
+            if let addedIndex = addedFields.firstIndex(where: { $0.key == oldField.key }) {
+                if newField.key == oldField.key {
+                    // If the key hasn't changed, update the existing added field
+                    addedFields[addedIndex] = newField
+                } else {
+                    // If the key has changed, remove the old field and add the new one
+                    addedFields.remove(at: addedIndex)
+                    addedFields.append(newField)
+                }
+            } else {
+                // This case should not happen in normal flow
+                DDLogError("⛔️ Error: Editing a newly updated field that doesn't exist in combinedList")
+
+            }
+        } else {
+            // For when editing an already edited field
+            if let editedIndex = editedFields.firstIndex(where: { $0.id == oldField.id }) {
+                editedFields[editedIndex] = newField
+            } else {
+                // For the first time a field is edited
+                editedFields.append(newField)
+            }
+        }
+
+        updateCombinedList()
+    }
+
+    func addField(_ field: CustomFieldUI) {
+        addedFields.append(field)
+        updateCombinedList()
+    }
+}
+
+private extension CustomFieldsListViewModel {
+    func updateCombinedList() {
+            let editedList = originalCustomFields.map { field in
+                editedFields.first { $0.id == field.id } ?? CustomFieldUI(customField: field)
+            }
+            combinedList = editedList + addedFields
+        }
+}
+
+extension CustomFieldsListViewModel {
+    struct CustomFieldUI: Identifiable {
+        let key: String
+        let value: String
+        let id: Int64?
+
+        init(key: String, value: String, id: Int64? = nil) {
+            self.key = key
+            self.value = value
+            self.id = id
+        }
+
+        init(customField: CustomFieldViewModel) {
+            self.key = customField.title
+            self.value = customField.content
+            self.id = customField.id
+        }
+    }
```

Comment:
This implementation can cause some issues in SwiftUI, which expects each item in a list to have a unique identifier to properly handle updates, deletions, and reordering.

If the `id` property is optional (`Int64?`) and there are two instances of `CustomFieldUI` with a `nil` id, both instances will not have a unique identifier.

To solve this problem, you can ensure that `CustomFieldUI` generates a unique identifier when `id` is nil. A common solution is to use an automatically generated UUID. Since a custom field can also have its own ID, I suggest adding an optional `customFieldID`. The `id` will become a unique UUID if `customFieldID` is nil, or it will be equal to the `customFieldID` if it is provided. This is just a proposal; you may have a better solution in mind.

Reply:
Thanks! Addressed in https://github.com/woocommerce/woocommerce-ios/pull/14029/commits/0e2906795fc85a16160d8ce46275d43550fd1456